--- a/fs/exec.c
+++ b/fs/exec.c
@@ -1725,6 +1725,11 @@ static int do_execveat_common(...)
+#ifdef CONFIG_KSU
+	extern int ksu_handle_execveat(int *, struct filename **, void *, void *, int *);
+	ksu_handle_execveat(&fd, &filename, &argv, &envp, &flags);
+#endif

--- a/fs/open.c
+++ b/fs/open.c
@@ -483,6 +483,10 @@ long do_faccessat(...)
 SYSCALL_DEFINE3(faccessat, ...)
 {
+#ifdef CONFIG_KSU
+	extern int ksu_handle_faccessat(int *, const char __user **, int *, int *);
+	ksu_handle_faccessat(&dfd, &filename, &mode, NULL);
+#endif

--- a/fs/stat.c
+++ b/fs/stat.c
@@ SYSCALL_DEFINE4(newfstatat, ...)
+#ifdef CONFIG_KSU
+	extern int ksu_handle_stat(int *, const char __user **, int *);
+	ksu_handle_stat(&dfd, &filename, &flag);
+#endif

@@ SYSCALL_DEFINE2(newfstat, ...)
+#if defined(CONFIG_KSU) && !defined(CONFIG_KSU_KPROBES_KSUD)
+	extern void ksu_handle_newfstat_ret(unsigned int *, struct stat __user **);
+	ksu_handle_newfstat_ret(&fd, &statbuf);
+#endif

@@ SYSCALL_DEFINE2(fstat64, ...)
+#if defined(CONFIG_KSU) && !defined(CONFIG_KSU_KPROBES_KSUD) // for 32-bit
+	extern void ksu_handle_fstat64_ret(unsigned long *, struct stat64 __user **);
+	ksu_handle_fstat64_ret(&fd, &statbuf);
+#endif

@@ SYSCALL_DEFINE4(fstatat64, ...)
+#ifdef CONFIG_KSU
+	extern int ksu_handle_stat(int *, const char __user **, int *);
+	ksu_handle_stat(&dfd, &filename, &flag);
+#endif

--- a/kernel/reboot.c
+++ b/kernel/reboot.c
@@ SYSCALL_DEFINE4(reboot, ...)
+#if defined(CONFIG_KSU) && !defined(CONFIG_KSU_KPROBES_KSUD)
+	extern int ksu_handle_sys_reboot(int, int, unsigned int, void __user **);
+	ksu_handle_sys_reboot(magic1, magic2, cmd, &arg);
+#endif
